api/include/opentelemetry/metrics/sync_instruments.h

```diff
@@ -194,5 +248,85 @@ class UpDownCounter : public SynchronousInstrument
   }
 };
 
+#if OPENTELEMETRY_ABI_VERSION_NO >= 2
+
+/** An Gauge instrument that records value.
+ * @since ABI_VERSION 2
+ */
+
+template <class T>
+class Gauge : public SynchronousInstrument
+{
+public:
+  /**
+   * Record a value
+   *
+   * @param value The increment amount. MUST be non-negative.
+   */
+  virtual void Record(T value) noexcept = 0;
+
+  /**
+   * Record a value
+   *
+   * @param value The increment amount. MUST be non-negative.
+   * @param context The explicit context to associate with this measurement.
+   */
+  virtual void Record(T value, const context::Context &context) noexcept = 0;
+
+  /**
+   * Record a value with a set of attributes.
+   *
+   * @param value The increment amount. MUST be non-negative.
+   * @param attributes A set of attributes to associate with the value.
+   */
+
+  virtual void Record(T value, const common::KeyValueIterable &attributes) noexcept = 0;
+
+  /**
+   * Record a value with a set of attributes.
+   *
+   * @param value The increment amount. MUST be non-negative.
+   * @param attributes A set of attributes to associate with the value.
+   * @param context The explicit context to associate with this measurement.
+   */
+  virtual void Record(T value,
+                      const common::KeyValueIterable &attributes,
+                      const context::Context &context) noexcept = 0;
```

Comment:
Here there are 4 virtual methods for the ABI, with different parameters combinations.

An alternative is to define only one virtual method for the ABI, accepting pointers instead of references:

```
  virtual void Record(T value,
                      const common::KeyValueIterable *attributes,
                      const context::Context *context) noexcept = 0;
```

and then have all the API helpers invoke this unique entry point, passing `nullptr` for optional parameters.

If needed, we can still have a helper like:

  void Record(T value,
                      const common::KeyValueIterable &attributes,
                      const context::Context &context) {
    this->Record(value, &attributes, &context);
  }

so that user code pass references as usual.

Making a clear distinction between the ABI (the virtual method) and the API helpers  (non virtual) will help,
because it is then easier to decide if a change is a breaking change or not.

Changes to the unique ABI virtual method is a breaking change.

Changes to the plenty of API helpers accepting combinations of parameters is not breaking the ABI,
as it always forwards to the same prototype.

Reply:
A much cleaner method! Thank you for the recommendation. I saw this approach in the GetMeter() PR but overlooked adding it here. :)